### PR TITLE
[IOTDB-2633]Unification of compaction log formats

### PR DESCRIPTION
**Pull request by choubenson**
1. Unify the log formats for inner space and cross space compaction. Fix new log compatible with 0.12
2. Handle when target file not exist while all source files exist in compaction recover.
3. Fix inner space compaction recovery from 0.12 when merge.mods exist.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java [hunk 1]
```diff
@@ -43,11 +43,11 @@
   public static final int LOGICAL_SG_OFFSET_IN_PATH = 4;
   public static final int SEQUENCE_OFFSET_IN_PATH = 5;
 
-  public static final int LOGICAL_SG_OFFSET_IN_LOG = 0;
-  public static final int VIRTUAL_SG_OFFSET_IN_LOG = 1;
-  public static final int TIME_PARTITION_OFFSET_IN_LOG = 2;
-  public static final int FILE_NAME_OFFSET_IN_LOG = 3;
-  public static final int SEQUENCE_OFFSET_IN_LOG = 4;
+  public static final int SEQUENCE_OFFSET_IN_LOG = 0;
```
Comment (xingtanzjr) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java:46]: After these offsets are changed, the sample in the comments of method `getFileIdentifierFromInfoString` is not correct

**Comment by xingtanzjr** [comment]
> 1. Unify the log formats for inner space and cross space compaction. Fix new log compatible with 0.12
> 2. Handle when target file not exist while all source files exist in compaction recover.
> 3. Fix recover inner space compaction from 0.12 when merge.mods exist.

A kindly suggestion... Let's use correct English grammar...

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java [hunk 2]
```diff
@@ -30,47 +29,41 @@
 import java.util.List;
 
 /** MergeLogger records the progress of a merge in file "merge.log" as text lines. */
-public class RewriteCrossSpaceCompactionLogger implements AutoCloseable {
+public class CompactionLogger implements AutoCloseable {
 
-  public static final String COMPACTION_LOG_NAME = "cross-compaction.log";
-  public static final String COMPACTION_LOG_NAME_FEOM_OLD = "merge.log";
+  public static final String CROSS_COMPACTION_LOG_NAME = "cross-compaction.log";
+  public static final String CROSS_COMPACTION_LOG_NAME_FEOM_OLD = "merge.log";
```
Comment (xingtanzjr) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java:35]: should be `CROSS_COMPACTION_LOG_NAME_FROM_OLD`

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java [hunk 3]
```diff
@@ -30,47 +29,41 @@
...
+  public static final String INNER_COMPACTION_LOG_NAME = ".compaction.log";
```
Comment (xingtanzjr) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java:36]: From my understanding(I am not very familiar with Merge now...), the value of `INNER_COMPACTION_LOG_NAME ` should be something like "inner-compaction.log" if we change the value of `CROSS_COMPACTION_LOG_NAME ` to "cross-compaction.log" ?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 4]
```diff
@@ -98,11 +101,15 @@ public void doCompaction() {
             logicalStorageGroupName,
             virtualStorageGroup,
             compactionLogFile);
-        SizeTieredCompactionLogAnalyzer logAnalyzer =
-            new SizeTieredCompactionLogAnalyzer(compactionLogFile);
-        logAnalyzer.analyze();
+        CompactionLogAnalyzer logAnalyzer = new CompactionLogAnalyzer(compactionLogFile);
+        if (compactionLogFile.getName().startsWith(tsFileManager.getStorageGroupName())) {
```
Comment (xingtanzjr) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:105]: It's better to define a method such as `isOldLogFile()` or `isLogFileVersionLowerThanZeroOneThree()` here and implement the details logic inside the method

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java (hunk 1 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java:46]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java [hunk 5]
```diff
@@ -126,26 +126,27 @@ public static TsFileIdentifier getFileIdentifierFromOldInfoString(String oldInfo
...
-        splittedFileInfo[LOGICAL_SG_OFFSET_IN_LOG],
-        splittedFileInfo[VIRTUAL_SG_OFFSET_IN_LOG],
-        splittedFileInfo[TIME_PARTITION_OFFSET_IN_LOG],
-        Boolean.parseBoolean(splittedFileInfo[SEQUENCE_OFFSET_IN_LOG]),
-        splittedFileInfo[FILE_NAME_OFFSET_IN_LOG]);
+        splittedFileInfo[LOGICAL_SG_OFFSET_IN_LOG - 1],
+        splittedFileInfo[VIRTUAL_SG_OFFSET_IN_LOG - 1],
+        splittedFileInfo[TIME_PARTITION_OFFSET_IN_LOG - 1],
+        splittedFileInfo[SEQUENCE_OFFSET_IN_LOG + 4].equals("true")
+            || splittedFileInfo[SEQUENCE_OFFSET_IN_LOG + 4].equals("sequence"),
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java:133]: Yes, infoString of old inner compaction log is sequence, info string of old cross compaction log is true. Resolved.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java (hunk 2 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java:35]: Resolved.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java (hunk 3 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/utils/log/CompactionLogger.java:36]: Resolved.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java (hunk 4 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:105]: Resolved.

**Review by jt2594838 — APPROVED** [review]

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java [hunk 6]
```diff
@@ -126,26 +128,27 @@ public static TsFileIdentifier getFileIdentifierFromOldInfoString(String oldInfo
...
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java [hunk 7]
```diff
@@ -134,23 +131,20 @@ private void executeCompaction()
...
-    try (RewriteCrossSpaceCompactionLogger compactionLogger =
-        new RewriteCrossSpaceCompactionLogger(logFile)) {
+    try (CompactionLogger compactionLogger = new CompactionLogger(logFile)) {
       // print the path of the temporary file first for priority check during recovery
+      compactionLogger.logFiles(selectedSeqTsFileResourceList, STR_SOURCE_FILES);
+      compactionLogger.logFiles(selectedUnSeqTsFileResourceList, STR_SOURCE_FILES);
       compactionLogger.logFiles(targetTsfileResourceList, STR_TARGET_FILES);
-      compactionLogger.logFiles(selectedSeqTsFileResourceList, STR_SEQ_FILES);
-      compactionLogger.logFiles(selectedUnSeqTsFileResourceList, STR_UNSEQ_FILES);
       CompactionUtils.compact(
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java:141]: may be we can close the logger before `compact`?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java [hunk 8]
```diff
@@ -104,16 +101,14 @@ protected void doCompaction() throws Exception {
...
-      }
-      sizeTieredCompactionLogger.logSequence(sequence);
-      sizeTieredCompactionLogger.logFileInfo(TARGET_INFO, targetTsFileResource.getTsFile());
+      sizeTieredCompactionLogger = new CompactionLogger(logFile);
+      sizeTieredCompactionLogger.logFiles(
+          selectedTsFileResourceList, CompactionLogger.STR_SOURCE_FILES);
+      sizeTieredCompactionLogger.logFiles(
+          Collections.singletonList(targetTsFileResource), CompactionLogger.STR_TARGET_FILES);
       LOGGER.info(
           "{} [Compaction] compaction with {}", fullStorageGroupName, selectedTsFileResourceList);
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java:113]: may be we can close the logger here

File: server/src/test/java/org/apache/iotdb/db/engine/compaction/recover/RewriteCrossSpaceCompactionRecoverCompatibleTest.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
+/*
```
Comment (THUMarkLau) [inline server/src/test/java/org/apache/iotdb/db/engine/compaction/recover/RewriteCrossSpaceCompactionRecoverCompatibleTest.java:1]: maybe we should test with log that use absolute path to locate files

**Review by THUMarkLau — CHANGES_REQUESTED** [review]
change according to comment

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java (hunk 7 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java (hunk 8 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java:115]: Resolved

File: server/src/test/java/org/apache/iotdb/db/engine/compaction/recover/RewriteCrossSpaceCompactionRecoverCompatibleTest.java (hunk 9 above)
Comment (choubenson) [reply server/src/test/java/org/apache/iotdb/db/engine/compaction/recover/RewriteCrossSpaceCompactionRecoverCompatibleTest.java:1]: Resolved

**Pull request merged**